File: 3/js/src/KnapsackProblem.ts
function getItems(memo: number[][], items: [number,number][]): [number, number][] {
	const res: [number, number][] = [];

	let i=memo.length-1, j=memo[0].length-1;
	while(i > 0 && j > 0) {
		if(memo[i][j] != memo[i-1][j]) {
			res.push(items[i-1]);
			j -= items[i-1][1];
		}

		i--;
	}

	return res;
}

function knapsackProblem(items: [number, number][], size: number): [number, number][] {
	items.sort( (a,b) => a[1] - b[1] );
	const memo: number[][] = [];

	for(let i=0; i<=items.length; i++) {
		const row: number[] = [];
		for(let j=0; j<=size; j++) {
			row.push(0);
		}
		memo.push(row);
	}

	for(let i=1; i<=items.length; i++) {
		const item = items[i-1];

		for(let j=0; j<item[1]; j++) {
			memo[i][j] = memo[i-1][j];
		}

		for(let j=item[1]; j<=size; j++) {
			memo[i][j] = item[0] + memo[i-1][j - item[1]];
		}
	}

	return getItems(memo, items);
}

(function main() {
	const items: [number, number][] = [[1,2], [4,3], [5,6], [6,7]];
	const size: number = 10;

	console.log(knapsackProblem(items, size));
})();

File: 3/java/KnapsackProblem.java
import java.util.ArrayList;
import java.util.List;

class KnapsackProblem {
		
	public static List<int[]> backtrack(int values[][], int items[][]) {
		int i = values.length-1;
		int j =  values[0].length-1;

		List<int[]> res = new ArrayList<>();
		
		while(i>=1 && j>=1) {
			int curItem[] = items[i-1];

			if(values[i][j] != values[i-1][j]) {
				res.add(curItem);
				i = i-1;
				j = j-curItem[1];
			} else {
				i = i-1;
			}	
		}

		return res;
	}

	public static List<int[]> solveKnapsackProblem(int items[][], int maxWeight) {
		int values[][] = new int[items.length+1][maxWeight+1];
	
		for(int i=1; i<=items.length; i++) {
			for(int j=1; j<=maxWeight; j++) {
				int curItem[] = items[i-1]; 
				if(curItem[1] <= j) {
					values[i][j] = Math.max( values[i-1][j], values[i-1][j-curItem[1]] + curItem[0] );
				} else {
					values[i][j] = values[i-1][j];
				}
			}
		}
		
		List<int[]> res = backtrack(values, items);	

		return res;
	}

	public static void main(String args[]) {
		int items[][] = {{1,2}, {4,3}, {5,6}, {6,7}};
		int maxWeight = 10;
		
		List<int[]> res = solveKnapsackProblem(items, maxWeight);
		
		for(int[] pair : res) {
			System.out.println(String.format("Value: %d, Weight: %d", pair[0], pair[1]));
		}
	}
}
